# Titanic competition with TensorFlow Decision Forests

This notebook will take you through the steps needed to train a baseline Gradient Boosted Trees Model using TensorFlow Decision Forests and creating a submission on the Titanic competition. 

This notebook shows:

1. How to do some basic pre-processing. For example, the passenger names will be tokenized, and ticket names will be splitted in parts.
1. How to train a Gradient Boosted Trees (GBT) with default parameters
1. How to train a GBT with improved default parameters
1. How to tune the parameters of a GBTs
1. How to train and ensemble many GBTs

# Imports dependencies

# 

In [1]:
import numpy as np
import pandas as pd
import os

import tensorflow as tf
import tensorflow_decision_forests as tfdf

print(f"Found TF-DF {tfdf.__version__}")

Found TF-DF 1.2.0


# Load dataset

In [2]:
train_df = pd.read_csv("/kaggle/input/titanic/train.csv")
serving_df = pd.read_csv("/kaggle/input/titanic/test.csv")

train_df.head(10)

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S
5,6,0,3,"Moran, Mr. James",male,NaN,0,0,330877,8.4583,NaN,Q
6,7,0,1,"McCarthy, Mr. Timothy J",male,54.0,0,0,17463,51.8625,E46,S
7,8,0,3,"Palsson, Master. Gosta Leonard",male,2.0,3,1,349909,21.0750,NaN,S
8,9,1,3,"Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)",female,27.0,0,2,347742,11.1333,NaN,S
9,10,1,2,"Nasser, Mrs. Nicholas (Adele Achem)",female,14.0,1,0,237736,30.0708,NaN,C


# Prepare dataset

We will apply the following transformations on the dataset.

1. Tokenize the names. For example, "Braund, Mr. Owen Harris" will become ["Braund", "Mr.", "Owen", "Harris"].
2. Extract any prefix in the ticket. For example ticket "STON/O2. 3101282" will become "STON/O2." and 3101282.

In [3]:
def preprocess(df):
    df = df.copy()
    
    def normalize_name(x):
        return " ".join([v.strip(",()[].\"'") for v in x.split(" ")])
    
    def ticket_number(x):
        return x.split(" ")[-1]
        
    def ticket_item(x):
        items = x.split(" ")
        if len(items) == 1:
            return "NONE"
        return "_".join(items[0:-1])
    
    df["Name"] = df["Name"].apply(normalize_name)
    df["Ticket_number"] = df["Ticket"].apply(ticket_number)
    df["Ticket_item"] = df["Ticket"].apply(ticket_item)                     
    return df
    
preprocessed_train_df = preprocess(train_df)
preprocessed_serving_df = preprocess(serving_df)

preprocessed_train_df.head(5)

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked,Ticket_number,Ticket_item
0,1,0,3,Braund Mr Owen Harris,male,22.0,1,0,A/5 21171,7.2500,NaN,S,21171,A/5
1,2,1,1,Cumings Mrs John Bradley Florence Briggs Thayer,female,38.0,1,0,PC 17599,71.2833,C85,C,17599,PC
2,3,1,3,Heikkinen Miss Laina,female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S,3101282,STON/O2.
3,4,1,1,Futrelle Mrs Jacques Heath Lily May Peel,female,35.0,1,0,113803,53.1000,C123,S,113803,NONE
4,5,0,3,Allen Mr William Henry,male,35.0,0,0,373450,8.0500,NaN,S,373450,NONE


Let's keep the list of the input features of the model. Notably, we don't want to train our model on the "PassengerId" and "Ticket" features.

In [4]:
input_features = list(preprocessed_train_df.columns)
input_features.remove("Ticket")
input_features.remove("PassengerId")
input_features.remove("Survived")
#input_features.remove("Ticket_number")

print(f"Input features: {input_features}")

Input features: ['Pclass', 'Name', 'Sex', 'Age', 'SibSp', 'Parch', 'Fare', 'Cabin', 'Embarked', 'Ticket_number', 'Ticket_item']


# Convert Pandas dataset to TensorFlow Dataset

In [5]:
def tokenize_names(features, labels=None):
    """Divite the names into tokens. TF-DF can consume text tokens natively."""
    features["Name"] =  tf.strings.split(features["Name"])
    return features, labels

train_ds = tfdf.keras.pd_dataframe_to_tf_dataset(preprocessed_train_df,label="Survived").map(tokenize_names)
serving_ds = tfdf.keras.pd_dataframe_to_tf_dataset(preprocessed_serving_df).map(tokenize_names)

# Train model with default parameters

### Train model

First, we are training a GradientBoostedTreesModel model with the default parameters.

In [6]:
model = tfdf.keras.GradientBoostedTreesModel(
    verbose=0, # Very few logs
    features=[tfdf.keras.FeatureUsage(name=n) for n in input_features],
    exclude_non_specified_features=True, # Only use the features in "features"
    random_seed=1234,
)
model.fit(train_ds)

self_evaluation = model.make_inspector().evaluation()
print(f"Accuracy: {self_evaluation.accuracy} Loss:{self_evaluation.loss}")

[INFO 2026-03-26T10:07:47.714907738+00:00 kernel.cc:1214] Loading model from path /tmp/tmp47ujyigj/model/ with prefix c76cb9942efe4520
[INFO 2026-03-26T10:07:47.722888558+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2026-03-26T10:07:47.72296135+00:00 kernel.cc:1046] Use fast generic engine


Please report this to the TensorFlow team. When filing the bug, set the verbosity to 10 (on Linux, `export AUTOGRAPH_VERBOSITY=10`) and attach the full output.
Cause: could not get source code
To silence this warning, decorate the function with @tf.autograph.experimental.do_not_convert
Accuracy: 0.8260869383811951 Loss:0.8608942627906799


# Train model with improved default parameters

Now you'll use some specific parameters when creating the GBT model

In [7]:
model = tfdf.keras.GradientBoostedTreesModel(
    verbose=0, # Very few logs
    features=[tfdf.keras.FeatureUsage(name=n) for n in input_features],
    exclude_non_specified_features=True, # Only use the features in "features"
    
    #num_trees=2000,
    
    # Only for GBT.
    # A bit slower, but great to understand the model.
    # compute_permutation_variable_importance=True,
    
    # Change the default hyper-parameters
    # hyperparameter_template="benchmark_rank1@v1",
    
    #num_trees=1000,
    #tuner=tuner
    
    min_examples=1,
    categorical_algorithm="RANDOM",
    #max_depth=4,
    shrinkage=0.05,
    #num_candidate_attributes_ratio=0.2,
    split_axis="SPARSE_OBLIQUE",
    sparse_oblique_normalization="MIN_MAX",
    sparse_oblique_num_projections_exponent=2.0,
    num_trees=2000,
    #validation_ratio=0.0,
    random_seed=1234,
    
)
model.fit(train_ds)

self_evaluation = model.make_inspector().evaluation()
print(f"Accuracy: {self_evaluation.accuracy} Loss:{self_evaluation.loss}")

[INFO 2026-03-26T10:07:50.500074418+00:00 kernel.cc:1214] Loading model from path /tmp/tmpqvdtnb46/model/ with prefix 565595b4125e491a
[INFO 2026-03-26T10:07:50.508519993+00:00 decision_forest.cc:661] Model loaded with 33 root(s), 1823 node(s), and 10 input feature(s).
[INFO 2026-03-26T10:07:50.508580201+00:00 kernel.cc:1046] Use fast generic engine


Accuracy: 0.760869562625885 Loss:1.0154211521148682


Let's look at the model and you can also notice the information about variable importance that the model figured out

In [8]:
model.summary()

Model: "gradient_boosted_trees_model_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
Total params: 1
Trainable params: 0
Non-trainable params: 1
_________________________________________________________________
Type: "GRADIENT_BOOSTED_TREES"
Task: CLASSIFICATION
Label: "__LABEL"

Input Features (11):
	Age
	Cabin
	Embarked
	Fare
	Name
	Parch
	Pclass
	Sex
	SibSp
	Ticket_item
	Ticket_number

No weights

Variable Importance: INV_MEAN_MIN_DEPTH:
    1.           "Sex"  0.576632 ################
    2.           "Age"  0.364297 #######
    3.          "Fare"  0.278839 ####
    4.          "Name"  0.208548 #
    5. "Ticket_number"  0.180792 
    6.        "Pclass"  0.176962 
    7.         "Parch"  0.176659 
    8.   "Ticket_item"  0.175540 
    9.      "Embarked"  0.172339 
   10.         "SibSp"  0.170442 

Variable Importance: NUM_AS_ROOT:
    1.  "Sex" 28.000000 ################
    2. "Name"  5.000000 

# Make predictions

In [9]:
def prediction_to_kaggle_format(model, threshold=0.5):
    proba_survive = model.predict(serving_ds, verbose=0)[:,0]
    return pd.DataFrame({
        "PassengerId": serving_df["PassengerId"],
        "Survived": (proba_survive >= threshold).astype(int)
    })

def make_submission(kaggle_predictions):
    path="/kaggle/working/submission.csv"
    kaggle_predictions.to_csv(path, index=False)
    print(f"Submission exported to {path}")
    
kaggle_predictions = prediction_to_kaggle_format(model)
make_submission(kaggle_predictions)
!head /kaggle/working/submission.csv

Submission exported to /kaggle/working/submission.csv
PassengerId,Survived
892,0
893,0
894,0
895,0
896,0
897,0
898,0
899,0
900,1


# Training a model with hyperparameter tunning

Hyper-parameter tuning is enabled by specifying the tuner constructor argument of the model. The tuner object contains all the configuration of the tuner (search space, optimizer, trial and objective).


In [10]:
tuner = tfdf.tuner.RandomSearch(num_trials=1000)
tuner.choice("min_examples", [2, 5, 7, 10])
tuner.choice("categorical_algorithm", ["CART", "RANDOM"])

local_search_space = tuner.choice("growing_strategy", ["LOCAL"])
local_search_space.choice("max_depth", [3, 4, 5, 6, 8])

global_search_space = tuner.choice("growing_strategy", ["BEST_FIRST_GLOBAL"], merge=True)
global_search_space.choice("max_num_nodes", [16, 32, 64, 128, 256])

#tuner.choice("use_hessian_gain", [True, False])
tuner.choice("shrinkage", [0.02, 0.05, 0.10, 0.15])
tuner.choice("num_candidate_attributes_ratio", [0.2, 0.5, 0.9, 1.0])


tuner.choice("split_axis", ["AXIS_ALIGNED"])
oblique_space = tuner.choice("split_axis", ["SPARSE_OBLIQUE"], merge=True)
oblique_space.choice("sparse_oblique_normalization",
                     ["NONE", "STANDARD_DEVIATION", "MIN_MAX"])
oblique_space.choice("sparse_oblique_weights", ["BINARY", "CONTINUOUS"])
oblique_space.choice("sparse_oblique_num_projections_exponent", [1.0, 1.5])

# Tune the model. Notice the `tuner=tuner`.
tuned_model = tfdf.keras.GradientBoostedTreesModel(tuner=tuner)
tuned_model.fit(train_ds, verbose=0)

tuned_self_evaluation = tuned_model.make_inspector().evaluation()
print(f"Accuracy: {tuned_self_evaluation.accuracy} Loss:{tuned_self_evaluation.loss}")

Use /tmp/tmpr2nb5n2e as temporary training directory


[INFO 2026-03-26T10:09:33.458323691+00:00 kernel.cc:1214] Loading model from path /tmp/tmpr2nb5n2e/model/ with prefix 23dd3a60a91b4c79
[INFO 2026-03-26T10:09:33.470433165+00:00 decision_forest.cc:661] Model loaded with 19 root(s), 589 node(s), and 12 input feature(s).
[INFO 2026-03-26T10:09:33.470467702+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesGeneric" built
[INFO 2026-03-26T10:09:33.470490594+00:00 kernel.cc:1046] Use fast generic engine


Accuracy: 0.9178082346916199 Loss:0.6503586769104004


In the last line in the cell above, you can see the accuracy is higher than previously with default parameters and parameters set by hand.

This is the main idea behing hyperparameter tuning.

For more information you can follow this tutorial: [Automated hyper-parameter tuning](https://www.tensorflow.org/decision_forests/tutorials/automatic_tuning_colab)

# Making an ensemble

Here you'll create 100 models with different seeds and combine their results

This approach removes a little bit the random aspects related to creating ML models

In the GBT creation is used the `honest` parameter. It will use different training examples to infer the structure and the leaf values. This regularization technique trades examples for bias estimates.

In [11]:
predictions = None
num_predictions = 0

for i in range(100):
    print(f"i:{i}")
    # Possible models: GradientBoostedTreesModel or RandomForestModel
    model = tfdf.keras.GradientBoostedTreesModel(
        verbose=0, # Very few logs
        features=[tfdf.keras.FeatureUsage(name=n) for n in input_features],
        exclude_non_specified_features=True, # Only use the features in "features"

        #min_examples=1,
        #categorical_algorithm="RANDOM",
        ##max_depth=4,
        #shrinkage=0.05,
        ##num_candidate_attributes_ratio=0.2,
        #split_axis="SPARSE_OBLIQUE",
        #sparse_oblique_normalization="MIN_MAX",
        #sparse_oblique_num_projections_exponent=2.0,
        #num_trees=2000,
        ##validation_ratio=0.0,
        random_seed=i,
        honest=True,
    )
    model.fit(train_ds)
    
    sub_predictions = model.predict(serving_ds, verbose=0)[:,0]
    if predictions is None:
        predictions = sub_predictions
    else:
        predictions += sub_predictions
    num_predictions += 1

predictions/=num_predictions

kaggle_predictions = pd.DataFrame({
        "PassengerId": serving_df["PassengerId"],
        "Survived": (predictions >= 0.5).astype(int)
    })

make_submission(kaggle_predictions)

i:0


[INFO 2026-03-26T10:09:34.262374018+00:00 kernel.cc:1214] Loading model from path /tmp/tmpfqofb9oq/model/ with prefix 909407c4e6254f13
[INFO 2026-03-26T10:09:34.266527188+00:00 kernel.cc:1046] Use fast generic engine


i:1


[INFO 2026-03-26T10:09:35.377763082+00:00 kernel.cc:1214] Loading model from path /tmp/tmp694i6qqt/model/ with prefix 0fb49533d28545bc
[INFO 2026-03-26T10:09:35.396561158+00:00 kernel.cc:1046] Use fast generic engine


i:2


[INFO 2026-03-26T10:09:36.205198526+00:00 kernel.cc:1214] Loading model from path /tmp/tmpntkhofxz/model/ with prefix b483a0853fdb4ac3
[INFO 2026-03-26T10:09:36.209715913+00:00 kernel.cc:1046] Use fast generic engine


i:3


[INFO 2026-03-26T10:09:37.505430376+00:00 kernel.cc:1214] Loading model from path /tmp/tmp64ygb90u/model/ with prefix c994a7a0b770413f
[INFO 2026-03-26T10:09:37.530434817+00:00 kernel.cc:1046] Use fast generic engine


i:4


[INFO 2026-03-26T10:09:38.347385866+00:00 kernel.cc:1214] Loading model from path /tmp/tmppv6aokic/model/ with prefix f3199539f8c14aa5
[INFO 2026-03-26T10:09:38.352777609+00:00 kernel.cc:1046] Use fast generic engine


i:5


[INFO 2026-03-26T10:09:39.157817124+00:00 kernel.cc:1214] Loading model from path /tmp/tmp9usgzcfk/model/ with prefix 7134706b0a0c4226
[INFO 2026-03-26T10:09:39.160956337+00:00 kernel.cc:1046] Use fast generic engine


i:6


[INFO 2026-03-26T10:09:40.031193168+00:00 kernel.cc:1214] Loading model from path /tmp/tmpl4ncq729/model/ with prefix 5ea0e4a013fb4e07
[INFO 2026-03-26T10:09:40.038477859+00:00 kernel.cc:1046] Use fast generic engine


i:7


[INFO 2026-03-26T10:09:41.250755949+00:00 kernel.cc:1214] Loading model from path /tmp/tmp301oa0_0/model/ with prefix 0f339f11df4a491e
[INFO 2026-03-26T10:09:41.270248721+00:00 kernel.cc:1046] Use fast generic engine


i:8


[INFO 2026-03-26T10:09:42.582160104+00:00 kernel.cc:1214] Loading model from path /tmp/tmpmt0y0pwv/model/ with prefix ba93519418d847c0
[INFO 2026-03-26T10:09:42.591662521+00:00 kernel.cc:1046] Use fast generic engine


i:9


[INFO 2026-03-26T10:09:43.658239363+00:00 kernel.cc:1214] Loading model from path /tmp/tmp5141agpn/model/ with prefix 3fec7afb1ec94ce5
[INFO 2026-03-26T10:09:43.672603899+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2026-03-26T10:09:43.672654898+00:00 kernel.cc:1046] Use fast generic engine


i:10


[INFO 2026-03-26T10:09:44.51297136+00:00 kernel.cc:1214] Loading model from path /tmp/tmp8tel_dx7/model/ with prefix 8804bb7ec51c4d27
[INFO 2026-03-26T10:09:44.518492815+00:00 kernel.cc:1046] Use fast generic engine


i:11


[INFO 2026-03-26T10:09:45.53226718+00:00 kernel.cc:1214] Loading model from path /tmp/tmpender65v/model/ with prefix 6c57a515e52b4bb2
[INFO 2026-03-26T10:09:45.546481163+00:00 kernel.cc:1046] Use fast generic engine


i:12


[INFO 2026-03-26T10:09:46.395572155+00:00 kernel.cc:1214] Loading model from path /tmp/tmp8dnsaqjp/model/ with prefix 59d95b4a0b6f4289
[INFO 2026-03-26T10:09:46.401120939+00:00 kernel.cc:1046] Use fast generic engine


i:13


[INFO 2026-03-26T10:09:47.372091559+00:00 kernel.cc:1214] Loading model from path /tmp/tmpjnym5bti/model/ with prefix c5419d5d61504699
[INFO 2026-03-26T10:09:47.382835371+00:00 kernel.cc:1046] Use fast generic engine


i:14


[INFO 2026-03-26T10:09:48.2114888+00:00 kernel.cc:1214] Loading model from path /tmp/tmpbw3ulegt/model/ with prefix 64e2f37bc64b4db0
[INFO 2026-03-26T10:09:48.217363923+00:00 kernel.cc:1046] Use fast generic engine


i:15


[INFO 2026-03-26T10:09:49.118973066+00:00 kernel.cc:1214] Loading model from path /tmp/tmpk0i0eeuk/model/ with prefix 25a4c425fe774fe7
[INFO 2026-03-26T10:09:49.12645743+00:00 kernel.cc:1046] Use fast generic engine


i:16


[INFO 2026-03-26T10:09:50.133692109+00:00 kernel.cc:1214] Loading model from path /tmp/tmps4d4yqyp/model/ with prefix c72f4af492aa4797
[INFO 2026-03-26T10:09:50.145944168+00:00 kernel.cc:1046] Use fast generic engine


i:17


[INFO 2026-03-26T10:09:51.146248227+00:00 kernel.cc:1214] Loading model from path /tmp/tmpuoa5x1st/model/ with prefix 6de9cd2a20b74cfd
[INFO 2026-03-26T10:09:51.158832139+00:00 kernel.cc:1046] Use fast generic engine


i:18


[INFO 2026-03-26T10:09:52.157051999+00:00 kernel.cc:1214] Loading model from path /tmp/tmp32f18e07/model/ with prefix 2ebc29f317e544ee
[INFO 2026-03-26T10:09:52.168202143+00:00 kernel.cc:1046] Use fast generic engine


i:19


[INFO 2026-03-26T10:09:53.318421732+00:00 kernel.cc:1214] Loading model from path /tmp/tmpq6hhu8kh/model/ with prefix 99f496b45b82415a
[INFO 2026-03-26T10:09:53.336142642+00:00 kernel.cc:1046] Use fast generic engine


i:20


[INFO 2026-03-26T10:09:54.407826146+00:00 kernel.cc:1214] Loading model from path /tmp/tmpj2hyt0z7/model/ with prefix 2024b4ac52e14ede
[INFO 2026-03-26T10:09:54.422738058+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2026-03-26T10:09:54.42278916+00:00 kernel.cc:1046] Use fast generic engine


i:21


[INFO 2026-03-26T10:09:55.251855676+00:00 kernel.cc:1214] Loading model from path /tmp/tmp555kij3i/model/ with prefix f7e857bf2fda484f
[INFO 2026-03-26T10:09:55.257035+00:00 kernel.cc:1046] Use fast generic engine


i:22


[INFO 2026-03-26T10:09:56.1160184+00:00 kernel.cc:1214] Loading model from path /tmp/tmp7j_j36e2/model/ with prefix 53e01bf7a8374875
[INFO 2026-03-26T10:09:56.121735834+00:00 kernel.cc:1046] Use fast generic engine


i:23


[INFO 2026-03-26T10:09:57.023708043+00:00 kernel.cc:1214] Loading model from path /tmp/tmpq4a50u3a/model/ with prefix d7bef84bb3574668
[INFO 2026-03-26T10:09:57.03202806+00:00 kernel.cc:1046] Use fast generic engine


i:24


[INFO 2026-03-26T10:09:57.89058984+00:00 kernel.cc:1214] Loading model from path /tmp/tmpwcvsnsgw/model/ with prefix 53eec27537e34c7a
[INFO 2026-03-26T10:09:57.896160725+00:00 kernel.cc:1046] Use fast generic engine


i:25


[INFO 2026-03-26T10:09:58.889625756+00:00 kernel.cc:1214] Loading model from path /tmp/tmpnibr7s51/model/ with prefix d281b675ab9d4c1e
[INFO 2026-03-26T10:09:58.900879765+00:00 kernel.cc:1046] Use fast generic engine


i:26


[INFO 2026-03-26T10:09:59.854210591+00:00 kernel.cc:1214] Loading model from path /tmp/tmpo0vb5ao9/model/ with prefix 2e431e0fd33a43cb
[INFO 2026-03-26T10:09:59.864088656+00:00 kernel.cc:1046] Use fast generic engine


i:27


[INFO 2026-03-26T10:10:00.723617239+00:00 kernel.cc:1214] Loading model from path /tmp/tmpyshf0omi/model/ with prefix 5c3329e519c44990
[INFO 2026-03-26T10:10:00.72985775+00:00 kernel.cc:1046] Use fast generic engine


i:28


[INFO 2026-03-26T10:10:01.581571266+00:00 kernel.cc:1214] Loading model from path /tmp/tmp09rqcwqu/model/ with prefix dc856d8e0ebb4a23
[INFO 2026-03-26T10:10:01.586951649+00:00 kernel.cc:1046] Use fast generic engine


i:29


[INFO 2026-03-26T10:10:03.216190632+00:00 kernel.cc:1214] Loading model from path /tmp/tmpad4wt4sk/model/ with prefix 04299ec334b94594
[INFO 2026-03-26T10:10:03.231221259+00:00 kernel.cc:1046] Use fast generic engine


i:30


[INFO 2026-03-26T10:10:04.57295635+00:00 kernel.cc:1214] Loading model from path /tmp/tmpgteu01xo/model/ with prefix 81c4629679db4d8a
[INFO 2026-03-26T10:10:04.600065162+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2026-03-26T10:10:04.600101856+00:00 kernel.cc:1046] Use fast generic engine


i:31


[INFO 2026-03-26T10:10:05.552483108+00:00 kernel.cc:1214] Loading model from path /tmp/tmp34owqh_9/model/ with prefix 60d4b4cb268046d4
[INFO 2026-03-26T10:10:05.561879405+00:00 kernel.cc:1046] Use fast generic engine


i:32


[INFO 2026-03-26T10:10:06.397851435+00:00 kernel.cc:1214] Loading model from path /tmp/tmp7v2lsb2z/model/ with prefix 7caa9462c82245b7
[INFO 2026-03-26T10:10:06.403641091+00:00 kernel.cc:1046] Use fast generic engine


i:33


[INFO 2026-03-26T10:10:07.397647929+00:00 kernel.cc:1214] Loading model from path /tmp/tmp4gy6x8cu/model/ with prefix 0f274b869d08490d
[INFO 2026-03-26T10:10:07.410441699+00:00 kernel.cc:1046] Use fast generic engine


i:34


[INFO 2026-03-26T10:10:08.342950969+00:00 kernel.cc:1214] Loading model from path /tmp/tmpv0bam2ik/model/ with prefix e576b6da237e4213
[INFO 2026-03-26T10:10:08.351380888+00:00 kernel.cc:1046] Use fast generic engine


i:35


[INFO 2026-03-26T10:10:09.245188445+00:00 kernel.cc:1214] Loading model from path /tmp/tmp6g984wk4/model/ with prefix f854cea713714602
[INFO 2026-03-26T10:10:09.252691393+00:00 kernel.cc:1046] Use fast generic engine


i:36


[INFO 2026-03-26T10:10:10.287740774+00:00 kernel.cc:1214] Loading model from path /tmp/tmp38c63k9p/model/ with prefix 5b1c58e547764f07
[INFO 2026-03-26T10:10:10.301901306+00:00 kernel.cc:1046] Use fast generic engine


i:37


[INFO 2026-03-26T10:10:11.215098413+00:00 kernel.cc:1214] Loading model from path /tmp/tmp70kykxtz/model/ with prefix f816047b9d5f40b9
[INFO 2026-03-26T10:10:11.223583906+00:00 kernel.cc:1046] Use fast generic engine


i:38


[INFO 2026-03-26T10:10:12.231023495+00:00 kernel.cc:1214] Loading model from path /tmp/tmpl1fr63_h/model/ with prefix 8be5d4b9dfdc4b72
[INFO 2026-03-26T10:10:12.245455317+00:00 kernel.cc:1046] Use fast generic engine


i:39


[INFO 2026-03-26T10:10:13.332934644+00:00 kernel.cc:1214] Loading model from path /tmp/tmpdgp9dh2d/model/ with prefix fbefbf1f78614a9d
[INFO 2026-03-26T10:10:13.345367544+00:00 kernel.cc:1046] Use fast generic engine


i:40


[INFO 2026-03-26T10:10:14.176434277+00:00 kernel.cc:1214] Loading model from path /tmp/tmpshl6y63j/model/ with prefix a385dd7f4b794d1b
[INFO 2026-03-26T10:10:14.180712273+00:00 kernel.cc:1046] Use fast generic engine


i:41


[INFO 2026-03-26T10:10:15.272527624+00:00 kernel.cc:1214] Loading model from path /tmp/tmpxkhiu0zt/model/ with prefix 25be9ade044f4056
[INFO 2026-03-26T10:10:15.287469255+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2026-03-26T10:10:15.287507047+00:00 kernel.cc:1046] Use fast generic engine


i:42


[INFO 2026-03-26T10:10:16.229366314+00:00 kernel.cc:1214] Loading model from path /tmp/tmpex6s6h9d/model/ with prefix f0eead2ff1eb43ed
[INFO 2026-03-26T10:10:16.238054324+00:00 kernel.cc:1046] Use fast generic engine


i:43


[INFO 2026-03-26T10:10:17.385105742+00:00 kernel.cc:1214] Loading model from path /tmp/tmpx635qp69/model/ with prefix ca8fbcdf68b849cd
[INFO 2026-03-26T10:10:17.402016972+00:00 kernel.cc:1046] Use fast generic engine


i:44


[INFO 2026-03-26T10:10:18.354080953+00:00 kernel.cc:1214] Loading model from path /tmp/tmpm5oonf2y/model/ with prefix 599fb85ef1034af1
[INFO 2026-03-26T10:10:18.363243212+00:00 kernel.cc:1046] Use fast generic engine


i:45


[INFO 2026-03-26T10:10:19.15078126+00:00 kernel.cc:1214] Loading model from path /tmp/tmpl9jxfmoa/model/ with prefix f263346346f64b32
[INFO 2026-03-26T10:10:19.154454581+00:00 kernel.cc:1046] Use fast generic engine


i:46


[INFO 2026-03-26T10:10:20.229016221+00:00 kernel.cc:1214] Loading model from path /tmp/tmpwyz728zp/model/ with prefix e6c604a0d9a240a8
[INFO 2026-03-26T10:10:20.242857208+00:00 kernel.cc:1046] Use fast generic engine


i:47


[INFO 2026-03-26T10:10:21.299033394+00:00 kernel.cc:1214] Loading model from path /tmp/tmp8su36knr/model/ with prefix ffb05e0243b04a67
[INFO 2026-03-26T10:10:21.311440118+00:00 kernel.cc:1046] Use fast generic engine


i:48


[INFO 2026-03-26T10:10:22.136061247+00:00 kernel.cc:1214] Loading model from path /tmp/tmp9ep2shrg/model/ with prefix 141d110f1f9f4b11
[INFO 2026-03-26T10:10:22.140604113+00:00 kernel.cc:1046] Use fast generic engine


i:49


[INFO 2026-03-26T10:10:23.089257422+00:00 kernel.cc:1214] Loading model from path /tmp/tmpjxryzpdo/model/ with prefix 4815faf1d8944d3e
[INFO 2026-03-26T10:10:23.095787174+00:00 kernel.cc:1046] Use fast generic engine


i:50


[INFO 2026-03-26T10:10:24.108705102+00:00 kernel.cc:1214] Loading model from path /tmp/tmpjy3qbpvi/model/ with prefix 2ba55f8cf12e4aee
[INFO 2026-03-26T10:10:24.120738255+00:00 kernel.cc:1046] Use fast generic engine


i:51


[INFO 2026-03-26T10:10:25.215707051+00:00 kernel.cc:1214] Loading model from path /tmp/tmpuqi7zej0/model/ with prefix 3ab3dbbd2c11453d
[INFO 2026-03-26T10:10:25.233102561+00:00 kernel.cc:1046] Use fast generic engine


i:52


[INFO 2026-03-26T10:10:26.134752812+00:00 kernel.cc:1214] Loading model from path /tmp/tmpt_3byi0m/model/ with prefix 188123a7bfd14b46
[INFO 2026-03-26T10:10:26.142825479+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2026-03-26T10:10:26.14285983+00:00 kernel.cc:1046] Use fast generic engine


i:53


[INFO 2026-03-26T10:10:27.654410893+00:00 kernel.cc:1214] Loading model from path /tmp/tmpxozxte2t/model/ with prefix 4df9cacfa01348cc
[INFO 2026-03-26T10:10:27.663463759+00:00 kernel.cc:1046] Use fast generic engine


i:54


[INFO 2026-03-26T10:10:28.547223568+00:00 kernel.cc:1214] Loading model from path /tmp/tmpc81ibirx/model/ with prefix 3b5ec2ba8d494acd
[INFO 2026-03-26T10:10:28.550927108+00:00 kernel.cc:1046] Use fast generic engine


i:55


[INFO 2026-03-26T10:10:29.648839446+00:00 kernel.cc:1214] Loading model from path /tmp/tmpyu7bc02e/model/ with prefix 938258ff4b6b4a49
[INFO 2026-03-26T10:10:29.664403827+00:00 kernel.cc:1046] Use fast generic engine


i:56


[INFO 2026-03-26T10:10:30.664383586+00:00 kernel.cc:1214] Loading model from path /tmp/tmpp2u0oaut/model/ with prefix 94f94b7d39da4c15
[INFO 2026-03-26T10:10:30.674974764+00:00 kernel.cc:1046] Use fast generic engine


i:57


[INFO 2026-03-26T10:10:31.561662017+00:00 kernel.cc:1214] Loading model from path /tmp/tmpcnlxhal9/model/ with prefix cc599f089c21411a
[INFO 2026-03-26T10:10:31.565976988+00:00 kernel.cc:1046] Use fast generic engine


i:58


[INFO 2026-03-26T10:10:32.465773875+00:00 kernel.cc:1214] Loading model from path /tmp/tmp28rep2l0/model/ with prefix 5731b441440f4610
[INFO 2026-03-26T10:10:32.4721088+00:00 kernel.cc:1046] Use fast generic engine


i:59


[INFO 2026-03-26T10:10:33.434740795+00:00 kernel.cc:1214] Loading model from path /tmp/tmp3eyt6eq1/model/ with prefix 9fbdae9a2b744eb9
[INFO 2026-03-26T10:10:33.443703977+00:00 kernel.cc:1046] Use fast generic engine


i:60


[INFO 2026-03-26T10:10:34.401951906+00:00 kernel.cc:1214] Loading model from path /tmp/tmp31yy_tjc/model/ with prefix ba28fbcf063a4383
[INFO 2026-03-26T10:10:34.411172203+00:00 kernel.cc:1046] Use fast generic engine


i:61


[INFO 2026-03-26T10:10:35.236305555+00:00 kernel.cc:1214] Loading model from path /tmp/tmp5xu51sbd/model/ with prefix b9ab6ca8339a498b
[INFO 2026-03-26T10:10:35.241009059+00:00 kernel.cc:1046] Use fast generic engine


i:62


[INFO 2026-03-26T10:10:36.53638713+00:00 kernel.cc:1214] Loading model from path /tmp/tmp5mqetc94/model/ with prefix 15013ed696b84102
[INFO 2026-03-26T10:10:36.560910803+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2026-03-26T10:10:36.560952642+00:00 kernel.cc:1046] Use fast generic engine


i:63


[INFO 2026-03-26T10:10:37.48088129+00:00 kernel.cc:1214] Loading model from path /tmp/tmpnr4i4o0w/model/ with prefix 6278d7753c3e4098
[INFO 2026-03-26T10:10:37.489440651+00:00 kernel.cc:1046] Use fast generic engine


i:64


[INFO 2026-03-26T10:10:38.354002292+00:00 kernel.cc:1214] Loading model from path /tmp/tmpjbkruivm/model/ with prefix 9d97d9a9ce4c4f2e
[INFO 2026-03-26T10:10:38.36148397+00:00 kernel.cc:1046] Use fast generic engine


i:65


[INFO 2026-03-26T10:10:39.182849373+00:00 kernel.cc:1214] Loading model from path /tmp/tmpltewn20a/model/ with prefix f803df3ed2164fd3
[INFO 2026-03-26T10:10:39.187801462+00:00 kernel.cc:1046] Use fast generic engine


i:66


[INFO 2026-03-26T10:10:40.078096517+00:00 kernel.cc:1214] Loading model from path /tmp/tmp46ql215b/model/ with prefix 0bd1fe5f71af4460
[INFO 2026-03-26T10:10:40.084233523+00:00 kernel.cc:1046] Use fast generic engine


i:67


[INFO 2026-03-26T10:10:41.252205626+00:00 kernel.cc:1214] Loading model from path /tmp/tmpo92d52ug/model/ with prefix be083d2b18154213
[INFO 2026-03-26T10:10:41.271452678+00:00 kernel.cc:1046] Use fast generic engine


i:68


[INFO 2026-03-26T10:10:42.273773466+00:00 kernel.cc:1214] Loading model from path /tmp/tmpwf8ect4f/model/ with prefix d2f8fb0804c3446a
[INFO 2026-03-26T10:10:42.285034299+00:00 kernel.cc:1046] Use fast generic engine


i:69


[INFO 2026-03-26T10:10:43.187801852+00:00 kernel.cc:1214] Loading model from path /tmp/tmpvijgjnes/model/ with prefix 9ea71d79e8f54c7a
[INFO 2026-03-26T10:10:43.194193815+00:00 kernel.cc:1046] Use fast generic engine


i:70


[INFO 2026-03-26T10:10:44.120876409+00:00 kernel.cc:1214] Loading model from path /tmp/tmpl7b8am75/model/ with prefix 6ae9011ea3e54911
[INFO 2026-03-26T10:10:44.129184558+00:00 kernel.cc:1046] Use fast generic engine


i:71


[INFO 2026-03-26T10:10:45.011694374+00:00 kernel.cc:1214] Loading model from path /tmp/tmp0b0ftmrd/model/ with prefix 52624d2ec4d4430e
[INFO 2026-03-26T10:10:45.019104916+00:00 kernel.cc:1046] Use fast generic engine


i:72


[INFO 2026-03-26T10:10:46.079217317+00:00 kernel.cc:1214] Loading model from path /tmp/tmp6ojnihaf/model/ with prefix 640abb21774e4341
[INFO 2026-03-26T10:10:46.094480723+00:00 kernel.cc:1046] Use fast generic engine


i:73


[INFO 2026-03-26T10:10:46.975614275+00:00 kernel.cc:1214] Loading model from path /tmp/tmps4hddskk/model/ with prefix 257c1b6fda7c4a12
[INFO 2026-03-26T10:10:46.982002734+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2026-03-26T10:10:46.982035702+00:00 kernel.cc:1046] Use fast generic engine


i:74


[INFO 2026-03-26T10:10:47.974879069+00:00 kernel.cc:1214] Loading model from path /tmp/tmpo1x3iy_c/model/ with prefix 018c68d6f93347cd
[INFO 2026-03-26T10:10:47.985808354+00:00 kernel.cc:1046] Use fast generic engine


i:75


[INFO 2026-03-26T10:10:48.899094823+00:00 kernel.cc:1214] Loading model from path /tmp/tmpp_npjdvm/model/ with prefix ccd0eb96c61048c8
[INFO 2026-03-26T10:10:48.907149471+00:00 kernel.cc:1046] Use fast generic engine


i:76


[INFO 2026-03-26T10:10:49.723450298+00:00 kernel.cc:1214] Loading model from path /tmp/tmpcxyib15i/model/ with prefix 8c9f8f526e234c39
[INFO 2026-03-26T10:10:49.727435093+00:00 kernel.cc:1046] Use fast generic engine


i:77


[INFO 2026-03-26T10:10:50.566254027+00:00 kernel.cc:1214] Loading model from path /tmp/tmp0kxf856k/model/ with prefix 19b7a5d020cb4adc
[INFO 2026-03-26T10:10:50.57073862+00:00 kernel.cc:1046] Use fast generic engine


i:78


[INFO 2026-03-26T10:10:51.444417887+00:00 kernel.cc:1214] Loading model from path /tmp/tmpdpu5uzha/model/ with prefix de8b1e332a1a4f65
[INFO 2026-03-26T10:10:51.450622793+00:00 kernel.cc:1046] Use fast generic engine


i:79


[INFO 2026-03-26T10:10:52.343610496+00:00 kernel.cc:1214] Loading model from path /tmp/tmp3zrfqrbv/model/ with prefix 7bc0c3fbf69d406d
[INFO 2026-03-26T10:10:52.350262435+00:00 kernel.cc:1046] Use fast generic engine


i:80


[INFO 2026-03-26T10:10:54.007825665+00:00 kernel.cc:1214] Loading model from path /tmp/tmp1wm19546/model/ with prefix 5b480e9f073d4316
[INFO 2026-03-26T10:10:54.019667207+00:00 kernel.cc:1046] Use fast generic engine


i:81


[INFO 2026-03-26T10:10:55.09998486+00:00 kernel.cc:1214] Loading model from path /tmp/tmppf8u81w2/model/ with prefix 136d9443a8754fd0
[INFO 2026-03-26T10:10:55.112538464+00:00 kernel.cc:1046] Use fast generic engine


i:82


[INFO 2026-03-26T10:10:56.10538296+00:00 kernel.cc:1214] Loading model from path /tmp/tmpvlk0pr_7/model/ with prefix 648aa26f83504c3c
[INFO 2026-03-26T10:10:56.115777549+00:00 kernel.cc:1046] Use fast generic engine


i:83


[INFO 2026-03-26T10:10:57.057691767+00:00 kernel.cc:1214] Loading model from path /tmp/tmpsrkr0eea/model/ with prefix 934eeea5947740f5
[INFO 2026-03-26T10:10:57.066295976+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2026-03-26T10:10:57.066328794+00:00 kernel.cc:1046] Use fast generic engine


i:84


[INFO 2026-03-26T10:10:58.151360506+00:00 kernel.cc:1214] Loading model from path /tmp/tmpe6djv5_2/model/ with prefix 2115942f0b7d4c54
[INFO 2026-03-26T10:10:58.167913496+00:00 kernel.cc:1046] Use fast generic engine


i:85


[INFO 2026-03-26T10:10:59.060179527+00:00 kernel.cc:1214] Loading model from path /tmp/tmplv59un9c/model/ with prefix fe155080f152446d
[INFO 2026-03-26T10:10:59.066517043+00:00 kernel.cc:1046] Use fast generic engine


i:86


[INFO 2026-03-26T10:11:00.13590452+00:00 kernel.cc:1214] Loading model from path /tmp/tmpnf7_3t12/model/ with prefix 720e4d2c1c8a4056
[INFO 2026-03-26T10:11:00.151604065+00:00 kernel.cc:1046] Use fast generic engine


i:87


[INFO 2026-03-26T10:11:01.317384111+00:00 kernel.cc:1214] Loading model from path /tmp/tmprisfvbe3/model/ with prefix 4a56e2fbeac64d1b
[INFO 2026-03-26T10:11:01.334115351+00:00 kernel.cc:1046] Use fast generic engine


i:88


[INFO 2026-03-26T10:11:02.318920977+00:00 kernel.cc:1214] Loading model from path /tmp/tmp28mgn9xe/model/ with prefix d99b65a26dcb4d9a
[INFO 2026-03-26T10:11:02.329710806+00:00 kernel.cc:1046] Use fast generic engine


i:89


[INFO 2026-03-26T10:11:03.196298535+00:00 kernel.cc:1214] Loading model from path /tmp/tmpj93ivr_0/model/ with prefix b4164c8e532f44bc
[INFO 2026-03-26T10:11:03.201194249+00:00 kernel.cc:1046] Use fast generic engine


i:90


[INFO 2026-03-26T10:11:04.168132987+00:00 kernel.cc:1214] Loading model from path /tmp/tmps_2rpace/model/ with prefix 8b009a437fff46ca
[INFO 2026-03-26T10:11:04.178107964+00:00 kernel.cc:1046] Use fast generic engine


i:91


[INFO 2026-03-26T10:11:05.059645575+00:00 kernel.cc:1214] Loading model from path /tmp/tmpabzb62qy/model/ with prefix 7e181febd1ea4d33
[INFO 2026-03-26T10:11:05.06638823+00:00 kernel.cc:1046] Use fast generic engine


i:92


[INFO 2026-03-26T10:11:06.172967555+00:00 kernel.cc:1214] Loading model from path /tmp/tmplap2q82e/model/ with prefix b4ec8369a7564654
[INFO 2026-03-26T10:11:06.189582029+00:00 kernel.cc:1046] Use fast generic engine


i:93


[INFO 2026-03-26T10:11:07.144880625+00:00 kernel.cc:1214] Loading model from path /tmp/tmpd1dsudqk/model/ with prefix f8cd79fa70834563
[INFO 2026-03-26T10:11:07.154634325+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2026-03-26T10:11:07.154667629+00:00 kernel.cc:1046] Use fast generic engine


i:94


[INFO 2026-03-26T10:11:08.045251418+00:00 kernel.cc:1214] Loading model from path /tmp/tmpgvy_dty9/model/ with prefix 36ff1283726943c5
[INFO 2026-03-26T10:11:08.051622522+00:00 kernel.cc:1046] Use fast generic engine


i:95


[INFO 2026-03-26T10:11:08.985395117+00:00 kernel.cc:1214] Loading model from path /tmp/tmp2e64_li7/model/ with prefix 1dca0113ac814c62
[INFO 2026-03-26T10:11:08.99355026+00:00 kernel.cc:1046] Use fast generic engine


i:96


[INFO 2026-03-26T10:11:09.947817873+00:00 kernel.cc:1214] Loading model from path /tmp/tmpp505dai0/model/ with prefix 463bb672c7cb44a5
[INFO 2026-03-26T10:11:09.957531106+00:00 kernel.cc:1046] Use fast generic engine


i:97


[INFO 2026-03-26T10:11:10.806516012+00:00 kernel.cc:1214] Loading model from path /tmp/tmpgv2p7rfc/model/ with prefix eea398c7cd8a4055
[INFO 2026-03-26T10:11:10.811504288+00:00 kernel.cc:1046] Use fast generic engine


i:98


[INFO 2026-03-26T10:11:11.733175523+00:00 kernel.cc:1214] Loading model from path /tmp/tmp5s7tgq7w/model/ with prefix 02ad1e46cbcd4edc
[INFO 2026-03-26T10:11:11.740473927+00:00 kernel.cc:1046] Use fast generic engine


i:99


[INFO 2026-03-26T10:11:12.82371936+00:00 kernel.cc:1214] Loading model from path /tmp/tmpmufmu7bz/model/ with prefix 34b4f6d78e3a49d9
[INFO 2026-03-26T10:11:12.837238286+00:00 kernel.cc:1046] Use fast generic engine


Submission exported to /kaggle/working/submission.csv


In [12]:
# ================================
# SIMPLE LOGISTIC REGRESSION
# ================================
import numpy as np
import pandas as pd

# Load data
train_df = pd.read_csv("/kaggle/input/titanic/train.csv")
test_df = pd.read_csv("/kaggle/input/titanic/train.csv")

# Preprocessing
def preprocess(df):
    df = df.copy()
    df["Age"].fillna(df["Age"].median(), inplace=True)
    df["Fare"].fillna(df["Fare"].median(), inplace=True)
    df["Embarked"].fillna("S", inplace=True)

    df["Sex"] = df["Sex"].map({"male":0, "female":1})
    df = pd.get_dummies(df, columns=["Embarked"], drop_first=True)

    return df

train_df = preprocess(train_df)
test_df = preprocess(test_df)

features = ["Pclass","Sex","Age","SibSp","Parch","Fare","Embarked_Q","Embarked_S"]

X = train_df[features].values
y = train_df["Survived"].values.reshape(-1,1)

X_test = test_df[features].values

# Sigmoid
def sigmoid(z):
    return 1 / (1 + np.exp(-z))

# Initialize
w = np.zeros((X.shape[1],1))
b = 0

# Training (manual gradient descent)
lr = 0.01
epochs = 1000

for i in range(epochs):
    z = X @ w + b
    y_hat = sigmoid(z)

    dz = y_hat - y
    dw = (1/len(X)) * (X.T @ dz)
    db = np.mean(dz)

    w -= lr * dw
    b -= lr * db

# Prediction
preds = sigmoid(X_test @ w + b)
preds = (preds >= 0.5).astype(int)

submission = pd.DataFrame({
    "PassengerId": test_df["PassengerId"],
    "Survived": preds.flatten()
})

submission.to_csv("/kaggle/working/submission_simple.csv", index=False)

In [13]:
# ================================
# JAX OPTIMIZED VERSION
# ================================
import jax
import jax.numpy as jnp
import pandas as pd

# Load + preprocess (same as before)
train_df = pd.read_csv("/kaggle/input/titanic/train.csv")
test_df = pd.read_csv("/kaggle/input/titanic/test.csv")

def preprocess(df):
    df = df.copy()
    df["Age"].fillna(df["Age"].median(), inplace=True)
    df["Fare"].fillna(df["Fare"].median(), inplace=True)
    df["Embarked"].fillna("S", inplace=True)

    df["Sex"] = df["Sex"].map({"male":0, "female":1})
    df = pd.get_dummies(df, columns=["Embarked"], drop_first=True)

    return df

train_df = preprocess(train_df)
test_df = preprocess(test_df)

features = ["Pclass","Sex","Age","SibSp","Parch","Fare","Embarked_Q","Embarked_S"]

X = jnp.array(train_df[features].values)
y = jnp.array(train_df["Survived"].values)

X_test = jnp.array(test_df[features].values)

# Model
def sigmoid(z):
    return 1 / (1 + jnp.exp(-z))

def predict(params, X):
    w, b = params
    return sigmoid(jnp.dot(X, w) + b)

# Loss
def loss_fn(params, X, y):
    preds = predict(params, X)
    return -jnp.mean(y * jnp.log(preds + 1e-8) + (1-y)*jnp.log(1-preds + 1e-8))

# Gradient
grad_fn = jax.grad(loss_fn)

# JIT compiled step
@jax.jit
def update(params, X, y, lr):
    grads = grad_fn(params, X, y)
    return (params[0] - lr * grads[0],
            params[1] - lr * grads[1])

# Initialize
params = (jnp.zeros(X.shape[1]), 0.0)

# Training
lr = 0.01
for i in range(1000):
    params = update(params, X, y, lr)

# Predictions (vmap for batch)
preds = predict(params, X_test)
preds = (preds >= 0.5).astype(int)

submission = pd.DataFrame({
    "PassengerId": test_df["PassengerId"],
    "Survived": np.array(preds)
})

submission.to_csv("/kaggle/working/submission_jax.csv", index=False)

In [14]:
# ============================================
# TITANIC: NORMAL vs JAX (jit + grad)
# ============================================

import numpy as np
import pandas as pd
import time

# JAX
import jax
import jax.numpy as jnp

# ================================
# LOAD DATA
# ================================
train_df = pd.read_csv("/kaggle/input/titanic/train.csv")
test_df = pd.read_csv("/kaggle/input/titanic/test.csv")

# ================================
# PREPROCESSING
# ================================
def preprocess(df):
    df = df.copy()

    df["Age"].fillna(df["Age"].median(), inplace=True)
    df["Fare"].fillna(df["Fare"].median(), inplace=True)
    df["Embarked"].fillna("S", inplace=True)

    df["Sex"] = df["Sex"].map({"male":0, "female":1})

    df = pd.get_dummies(df, columns=["Embarked"], drop_first=True)

    return df

train_df = preprocess(train_df)
test_df = preprocess(test_df)

features = ["Pclass","Sex","Age","SibSp","Parch","Fare","Embarked_Q","Embarked_S"]

X = train_df[features].values
y = train_df["Survived"].values.reshape(-1,1)

X_test = test_df[features].values

# ============================================
# 1. NORMAL LOGISTIC REGRESSION
# ============================================

def sigmoid_np(z):
    return 1 / (1 + np.exp(-z))

w = np.zeros((X.shape[1],1))
b = 0

lr = 0.01
epochs = 1000

start = time.time()

for i in range(epochs):
    z = X @ w + b
    y_hat = sigmoid_np(z)

    dz = y_hat - y
    dw = (1/len(X)) * (X.T @ dz)
    db = np.mean(dz)

    w -= lr * dw
    b -= lr * db

end = time.time()

normal_time = end - start

# Accuracy (train)
preds_train = (sigmoid_np(X @ w + b) >= 0.5).astype(int)
normal_acc = np.mean(preds_train == y)

# Test prediction
preds_test = (sigmoid_np(X_test @ w + b) >= 0.5).astype(int)

submission_normal = pd.DataFrame({
    "PassengerId": test_df["PassengerId"],
    "Survived": preds_test.flatten()
})

submission_normal.to_csv("/kaggle/working/submission_normal.csv", index=False)

# ============================================
# 2. JAX VERSION (jit + grad)
# ============================================

X_j = jnp.array(X)
y_j = jnp.array(y.flatten())
X_test_j = jnp.array(X_test)

def sigmoid_j(z):
    return 1 / (1 + jnp.exp(-z))

def predict(params, X):
    w, b = params
    return sigmoid_j(jnp.dot(X, w) + b)

def loss_fn(params, X, y):
    preds = predict(params, X)
    return -jnp.mean(y * jnp.log(preds + 1e-8) + (1-y)*jnp.log(1-preds + 1e-8))

grad_fn = jax.grad(loss_fn)

@jax.jit
def update(params, X, y, lr):
    grads = grad_fn(params, X, y)
    return (params[0] - lr * grads[0],
            params[1] - lr * grads[1])

params = (jnp.zeros(X_j.shape[1]), 0.0)

# 🔥 Warm-up (important!)
for i in range(10):
    params = update(params, X_j, y_j, lr)

start = time.time()

for i in range(epochs):
    params = update(params, X_j, y_j, lr)

end = time.time()

jax_time = end - start

# Accuracy (train)
jax_preds_train = (predict(params, X_j) >= 0.5).astype(int)
jax_acc = jnp.mean(jax_preds_train == y_j)

# Test prediction
jax_preds_test = (predict(params, X_test_j) >= 0.5).astype(int)

submission_jax = pd.DataFrame({
    "PassengerId": test_df["PassengerId"],
    "Survived": np.array(jax_preds_test)
})

submission_jax.to_csv("/kaggle/working/submission_jax.csv", index=False)

# ============================================
# 3. FINAL COMPARISON OUTPUT
# ============================================

print("\n========== FINAL COMPARISON ==========")
print(f"Normal Accuracy : {normal_acc:.4f}")
print(f"JAX Accuracy    : {float(jax_acc):.4f}")

print(f"\nNormal Time     : {normal_time:.4f} sec")
print(f"JAX Time        : {jax_time:.4f} sec")

print("\nSpeedup (Normal / JAX):", normal_time / jax_time)
print("======================================")


========== FINAL COMPARISON ==========
Normal Accuracy : 0.6813
JAX Accuracy    : 0.6162

Normal Time     : 0.0399 sec
JAX Time        : 0.1710 sec

Speedup (Normal / JAX): 0.23334104903841862


In [15]:
import platform
import os

print("OS:", platform.system(), platform.release())
print("Processor:", platform.processor())
print("CPU cores:", os.cpu_count())


OS: Linux 6.6.113+
Processor: x86_64
CPU cores: 4
